### ISSUE TITLE
Rarely calls to `produce_sync` will wait forever

### ISSUE THREAD

**Issue by sdball** [kafka4beam/brod#211]
I'm using brod in an Elixir production app that spins up a GenServer process per Kafka message that calls `:brod.produce_sync` to produce to Kafka. It works wonderfully the vast majority of the time but twice now in a couple months we've run into a state where the brod client does not return a response from `produce_sync`. This causes our worker processes to also hang forever since they're blocking on that function call and not even able to trigger an OTP timeout or respond to `:sys.get_status`.

The problem _seems_ to resolve if I call `:brod.stop_client(:brod_client_1)` and then restart it manually. I say seems because I've yet to truly isolate the condition and we've only had this failure state happen twice so far.

While the problem was in progress I did capture the status of the brod client if that's helpful

```erlang
iex> :sys.get_status(:brod_client_1)
{:status, #PID<0.756.0>, {:module, :gen_server},
 [["$initial_call": {:brod_client, :init, 1},
   "$ancestors": [:brod_sup, #PID<0.734.0>]], :running, #PID<0.735.0>, [],
  [header: 'Status for generic server brod_client_1',
   data: [{'Status', :running}, {'Parent', #PID<0.735.0>},
    {'Logged events', []}],
   data: [{'State',
     {:state, :brod_client_1,
      ["34.195.140.195": 9096, "34.195.140.72": 9096, "34.195.135.147": 9096],
      {:dead_since, {1495, 133506, 193413}, {:shutdown, :ssl_closed}},
      [{:sock, {'ec2-34-195-135-147.compute-1.amazonaws.com', 9096},
        #PID<0.824.0>},
       {:sock, {'ec2-34-195-140-195.compute-1.amazonaws.com', 9096},
        #PID<0.842.0>},
       {:sock, {'ec2-34-195-140-72.compute-1.amazonaws.com', 9096},
        #PID<0.853.0>}], #PID<0.779.0>, #PID<0.780.0>,
      [auto_start_producers: true,
       ssl: [certfile: "/etc/spreedly/ssl/heroku_kafka.crt",
        keyfile: "/etc/spreedly/ssl/heroku_kafka.key"],
       config: [restart_delay_seconds: 10, max_metadata_sock_retry: 1,
        get_metadata_timeout_seconds: 5, reconnect_cool_down_seconds: 1,
        allow_topic_auto_creation: false,
        default_producer_config: [required_acks: -1, ack_timeout: 1000,
         partition_buffer_limit: 65536, partition_onwire_limit: 1,
         max_batch_size: 1048576, max_retries: -1, retry_backoff_ms: 500,
         compression: :no_compression, min_compression_batch_size: 1024,
         topic_restart_delay_seconds: 1, partition_restart_delay_seconds: 3]]],
      :brod_client_1}}]]]}
```

Is this failure state something that brod could detect and respond to (e.g. restart the brod client)?

**Comment by zmstone** [comment]
there was a 'block firever' race fixed in 2.3.4
are you using a version earlier than that?

**Comment by sdball** [comment]
Ah I bet that's it! We're still on 2.2.15, thanks for the heads up! 🎉
